File: lab06b.py
#!/usr/bin/env python3

def is_safe(position, queens):
    """
    Test the safety of placing a queen in the given position
    taking into account the given positions of other queens.
    """
    (row, col) = position
    for (occupied_row, occupied_col) in queens:
        if (row == occupied_row
            or col == occupied_col
            or abs(row - occupied_row) == abs(col - occupied_col)):
            return False
    return True


def solve(n):
    """
    Attempt to place N queens on an NxN board.
    Place each queen in the first available safe position.
    Returns a list of up to N positions.
    """
    queens = list()
    for row in range(n):
        for col in range(n):
            position = (row, col)
            if is_safe(position, queens):
                queens.append(position)
                break # cannot place additional queens in same row
    return queens


def test():
    for n in range(1, 8):
        queens = solve(n)
        print('placed', len(queens), 'queens on', n, 'x', n, 'board')
        print(queens)

test()

File: src/bin/lab06b.rs
//! Programming Assignment 6: N queens

/// Demonstrate N queens solution.
fn main() {
    for size in 1..9 {
        let board = solve(size);
        println!("placed {} queens on {size} x {size} board", board.len());
        println!("{board:?}");
    }
}

/// A board is a collection of positions as (row, column) tuples.
type Board = Vec<(i32, i32)>;

/// Attempt to place N queens on an NxN board.
/// Return a collection of N positions upon success.
/// Return an empty collection upon failure.
fn solve(size: i32) -> Board {
    let mut board = Board::new();
    solve_recursively(&mut board, size, 0);
    return board;
}

/// Attempt to place a queen in the given row of an NxN board
/// taking into account the positions occupied by other queens.
/// Recursively attempt to place a queen in the next row
/// and backtrack when necessary.
/// Return true upon success.
fn solve_recursively(board: &mut Board, size: i32, row: i32) -> bool {
    for col in 0..size {
        let position = (row, col);
        if is_safe(board, position) {
            board.push(position); // tentatively place queen
            if row + 1 < size {
                if solve_recursively(board, size, row + 1) {
                    return true;
                } else {
                    board.pop(); // backtrack
                }
            } else {
                return true;
            }
        }
    }
    return false;
}

/// Test the safety of placing a queen in the given position
/// taking into account the positions occupied by other queens.
fn is_safe(board: &Board, position: (i32, i32)) -> bool {
    for occupied in board {
        if position.0 == occupied.0
            || position.1 == occupied.1
            || (position.0 - occupied.0).abs() == (position.1 - occupied.1).abs()
        {
            return false;
        }
    }
    return true;
}
